specs/BrowserProcessExited.md

```diff
@@ -0,0 +1,322 @@
+        /// the failed process.
+        Failed
+    };
+
+    runtimeclass CoreWebView2Environment
+    {
+        // ...
+
+        /// `BrowserProcessExited` is raised when the browser process of the
+        /// WebView2 Runtime associated to this `CoreWebView2Environment`
+        /// terminates due to an error or normal shutdown (e.g., when all its
+        /// WebViews are closed), after all resources (including the user data
+        /// folder) used by the browser process (and related processes) have
+        /// been released.
+        ///
+        /// All `CoreWebView2Environment` objects across different app processes that use
+        /// the same browser process receive this event when the browser process (and
+        /// associated processes) exits. If the browser process (and therefore the user data
+        /// folder) in use by the app process (through the `CoreWebView2Environment` options
+        /// used) is shared with other processes, these processes need to coordinate to
+        /// handle the potential race condition on the use of the resources. E.g., if one
+        /// app process tries to clear the user data folder, while other tries to recreate
+        /// its WebViews on crash.
+        ///
+        /// Note this is an event from the `ICoreWebView2Environment3` interface, not the
+        /// `ICoreWebView2`. The difference between this `BrowserProcessExited` event and
+        /// the `CoreWebView2`'s `ProcessFailed` event is that `BrowserProcessExited` is
+        /// raised for any (expected and unexpected) **browser process** (along its
+        /// associated processes) exits, while `ProcessFailed` is raised only for
+        /// **unexpected** browser process exits, or for **render process**
+        /// exits/unresponsiveness. To learn more about the WebView2 Process Model, go to
+        /// [Process model](https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/process-model).
+        ///
+        /// In the case the browser process crashes, both `BrowserProcessExited` and
+        /// `ProcessFailed` events are raised, but the order is not guaranteed.
+        event Windows.Foundation.TypedEventHandler<CoreWebView2Environment, CoreWebView2BrowserProcessExitedEventArgs> BrowserProcessExited;
+    }
+
+    /// Event args for the `CoreWebView2Environment.BrowserProcessExited` event.
+    runtimeclass CoreWebView2BrowserProcessExitedEventArgs
+    {
+        /// The kind of browser process exit that has occurred.
+        CoreWebView2BrowserProcessExitKind BrowserProcessExitKind { get; };
+    }
+}
+```
+
+# Appendix
+We expect that for the two scenarios this API is designed for, namely cleanup of
+the user data folder and upgrading the WebView2 Runtime, an app adding a
+handler for `BrowserProcessExited` will only be interested in the next single
+time the browser process exits (even if there could be more browser processes
```

Comment:
Addendum:
An additional consideration is the app registering the handler might be interested in a browser process exit that has already occurred (if the process terminates in between the environment connecting to the process and the API call).

For these cases, the registered handler will not be called, as the event already occurred. Invoking the handler if the process already exited would cover this, but wouldn't fit event behavior, and while the handler can be registered before the first WebView creation (which connects to the browser) to catch the event, the event API is not as easy to use as an async method would be.

Reply:
Recommendation: Its messy to have an event fire on subscription. For a single proc case there's no real problem - the host app can subscribe before the browser process gets created. For mult-proc host app, it is up to the app to coordinate between the host procs.